Replace str path utils with new `PathLookup` type

Alexendoo opened the pull request:
The `&[&str]` path based `clippy_utils` have been removed and replace with a new type `PathLookup`:

- [`match_trait_method`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_trait_method.html)
- [`match_qpath`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_qpath.html)
- [`match_path`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_path.html)
- [`match_any_def_paths`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_any_def_paths.html)
- [`match_def_path`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_def_path.html)
- [`match_type`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/ty/fn.match_type.html)
- [`get_trait_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.get_trait_def_id.html)

Internally `PathLookup` is a lazy call to `lookup_path` (the new name for [`def_path_res`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.def_path_res.html) to distinguish it from [`path_res`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.path_res.html))

The `invalid_paths` internal lint is removed, it could be reimplemented but it feels redundant since every path should be covered by a test anyway

### User facing changes

- `manual_saturating_arithmetic` now checks for `u32::MAX/MIN` instead of only detecting the legacy numeric consts (`std::u32::MAX/MIN`), `clippy::legacy_numeric_constants` will redirect usages of the legacy versions to the new one

- `allow-invalid = true` now suppresses all invalid path warnings, currently you can run into a warning that can't be ignored in some situations, e.g. with `serde` without the `derive` feature

  ```
  warning: expected a macro, found a trait
   --> /home/gh-Alexendoo/temp/clippy.toml:2:5
    |
  2 |     { path = "serde::Serialize", allow-invalid = true },
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```

- Re-exports of primitives types like `std::primitive::*` no longer work in `disallowed-types`, this seems acceptable since it would be unusual to deny a primitive this way rather than writing e.g. `usize`. Type aliases such as `c_char` are unaffected

- A similar slight performance improvement to https://github.com/rust-lang/rust-clippy/pull/14650
  ```bash
  $ hyperfine -w 2 -p 'touch src/cargo/lib.rs' 'cargo +master clippy' 'cargo +lazy-paths clippy'
  ```
  ```
  Benchmark 1: cargo +master clippy
    Time (mean ± σ):      6.829 s ±  0.064 s    [User: 6.079 s, System: 0.673 s]
    Range (min … max):    6.705 s …  6.907 s    10 runs
   
  Benchmark 2: cargo +lazy-paths clippy
    Time (mean ± σ):      6.765 s ±  0.064 s    [User: 5.984 s, System: 0.698 s]
    Range (min … max):    6.636 s …  6.834 s    10 runs
   
  Summary
    cargo +lazy-paths clippy ran
      1.01 ± 0.01 times faster than cargo +master clippy
  ```

changelog: none

Jarcho commented on book/src/development/trait_checking.md:
```diff
@@ -73,22 +73,24 @@ impl LateLintPass<'_> for CheckDropTraitLint {
 If neither diagnostic item nor a language item is available, we can use
-[`clippy_utils::paths`][paths] with the `match_trait_method` to determine trait
-implementation.
+[`clippy_utils::paths`][paths] to determine get a trait's `DefId`.
```
s/determine get/get

Jarcho commented on clippy_utils/src/paths.rs:
```diff
@@ -4,62 +4,104 @@
+/// - `std::vec` will return one [`DefId`] for the module and one for the macro
+pub struct PathLookup {
+    path: &'static [Symbol],
+    once: OnceLock<Vec<DefId>>,
```
This being a vec of mixed resolution types makes this very hard to use. There's no way to know which order the results are in.

Jarcho reviewed (commented):
This also breaks the author lint. It didn't really work properly before with `match_qpath`, but it's even more broken now.

Alexendoo commented:
- The paths now have to specify the namespace they want to resolve in, this also allowed `item_child_by_name` to no longer allocate
- Replaced `PathLookup::first` with `PathLookup::only`
- Added some suggestions for paths to external definitions for `clippy::author`

Jarcho commented on clippy_utils/src/lib.rs:
```diff
@@ -586,199 +513,205 @@ pub fn path_def_id<'tcx>(cx: &LateContext<'_>, maybe_path: &impl MaybePath<'tcx>
     };
 
-    let crates = find_primitive_impls(tcx, base)
-        .chain(local_crate)
-        .map(|id| Res::Def(tcx.def_kind(id), id))
-        .chain(find_crates(tcx, base_sym))
-        .collect();
-
-    def_path_res_with_base(tcx, crates, path)
+    let mut out = Vec::new();
+    for &base in find_crates(tcx, root).iter().chain(find_primitive_impls(tcx, root)) {
+        lookup_path_with_base(tcx, base, ns, rest, &mut out);
+    }
+    out
 }
```
Would it make sense to move this into the `paths` module? The crate root is kind of a mess right now.

Alexendoo replied:
Yeah good plan, moved them across

Jarcho reviewed (approved):
Definitely a lot nicer than what we had before. Just one question, but this looks good either way.
